django/core/cache/backends/db.py

```diff
@@ -49,52 +49,78 @@ class DatabaseCache(BaseDatabaseCache):
         connection = connections[db]
         quote_name = connection.ops.quote_name
         table = quote_name(self._table)
 
+        if not keys:
+            return dict()
+
         with connection.cursor() as cursor:
+            if len(key_map) == 1:
+                format_string = '= %s'
+            else:
+                format_string = 'IN(%s' + ', %s' * (len(key_map) - 1) + ')'
             cursor.execute(
-                'SELECT %s, %s, %s FROM %s WHERE %s = %%s' % (
+                'SELECT %s, %s, %s FROM %s WHERE %s %s' % (
                     quote_name('cache_key'),
                     quote_name('value'),
                     quote_name('expires'),
                     table,
                     quote_name('cache_key'),
+                    format_string,
                 ),
-                [key]
+                list(key_map)
             )
-            row = cursor.fetchone()
-        if row is None:
-            return default
+            rows = cursor.fetchall()
 
-        expires = row[2]
+        expired_keys = list()
+        return_dict = dict()
         expression = models.Expression(output_field=models.DateTimeField())
-        for converter in (connection.ops.get_db_converters(expression) +
-                          expression.get_db_converters(connection)):
-            if func_supports_parameter(converter, 'context'):  # RemovedInDjango30Warning
-                expires = converter(expires, expression, connection, {})
+        converters = (connection.ops.get_db_converters(expression) + expression.get_db_converters(connection))
+        for row in rows:
+            key = row[0]
+            value = row[1]
+            expires = row[2]
+
+            for converter in converters:
+                if func_supports_parameter(converter, 'context'):  # RemovedInDjango30Warning
+                    expires = converter(expires, expression, connection, {})
+                else:
+                    expires = converter(expires, expression, connection)
+
+            if expires < timezone.now():
+                expired_keys.append(key)
             else:
-                expires = converter(expires, expression, connection)
+                value = connection.ops.process_clob(value)
+                value = pickle.loads(base64.b64decode(value.encode()))
+                return_dict.update({
+                    key_map.get(key): value
+                })
 
-        if expires < timezone.now():
+        if expired_keys:
             db = router.db_for_write(self.cache_model_class)
             connection = connections[db]
             with connection.cursor() as cursor:
+                if len(expired_keys) == 1:
+                    format_string = '= %s'
+                else:
+                    format_string = 'IN(%s' + ', %s' * (len(expired_keys) - 1) + ')'
                 cursor.execute(
-                    'DELETE FROM %s WHERE %s = %%s' % (
+                    'DELETE FROM %s WHERE %s %s' % (
```

Comment:
This feels like a copy-paste of `delete_many()` that you have proposed in #10258.
Perhaps this could be rebased on top of that branch and all of this expiry code can condense to use:
```python
self.delete_many(self, expired_keys, version=version)
```